upstairs/src/dummy_downstairs_tests.rs

```diff
@@ -2646,3 +2646,57 @@ async fn test_write_replay() {
     // requires going to the worker thread.
     harness.guest.get_uuid().await.unwrap();
 }
+
+/// Test that messages do not make it to a Downstairs mid-negotiation
+#[tokio::test]
+async fn test_no_send_offline() {
+    let mut harness = TestHarness::new().await;
+    harness.ds1().cfg.reply_to_ping = false;
+
+    // Sleep until we're confident that the Downstairs is kicked out, answering
+    // pings from the other two Downstairs
+    info!(harness.log, "waiting for Upstairs to kick out DS1");
+    let sleep_time = Duration::from_secs_f32(CLIENT_TIMEOUT_SECS + 5.0);
+    tokio::select! {
+        e = harness.ds1.as_mut().unwrap().recv() => {
+            // We've been kicked out!
+            assert!(e.is_none());
+        }
+        _ = harness.ds2.recv() => panic!("unexpected message"),
+        _ = harness.ds3.recv() => panic!("unexpected message"),
+        _ = tokio::time::sleep(sleep_time) => panic!("unexpected timeout"),
+    }
+
+    // Check to make sure that happened
+    let ds = harness.guest.downstairs_state().await.unwrap();
+    assert_eq!(ds[ClientId::new(0)], DsState::Offline);
+    assert_eq!(ds[ClientId::new(1)], DsState::Active);
+    assert_eq!(ds[ClientId::new(2)], DsState::Active);
+    println!("WE'RE OFFLINE");
+
+    // Reconnect ds1
+    harness.restart_ds1().await;
+    println!("WE'RE RESTARTING");
+
+    // Wait for the Upstairs to try reconnecting, answering pings all the while
+    let reconnect_time = CLIENT_RECONNECT_DELAY + Duration::from_secs(2);
+    tokio::select! {
+        _ = harness.ds2.recv() => panic!("unexpected message"),
+        _ = harness.ds3.recv() => panic!("unexpected message"),
+        _ = tokio::time::sleep(reconnect_time) => (),
+    }
+
+    // Start negotiation
+    harness.ds1().negotiate_start().await;
+
+    // Check that a write doesn't make it to DS1
+    let write_handle = harness.spawn(|guest| async move {
+        let mut data = BytesMut::new();
+        data.resize(512, 1u8);
+        guest.write(BlockIndex(0), data).await.unwrap();
+    });
+    harness.ds2.ack_write().await;
+    harness.ds3.ack_write().await;
+    assert_eq!(harness.ds1().try_recv(), Err(TryRecvError::Empty));
+    write_handle.await.unwrap();
```

Comment:
Maybe extend this to bring ds1 online and expect the write is send?

Reply:
Good idea, done